### PR TITLE
✨ feat: new validator to check Multiple Architectures configuration

### PR DESCRIPTION
**Pull request by camilamacedo86**
**Description**
add: new validator to check Multiarch configuration

**Motivation**
Help users, cis and tools verify the  Multiarch configuration

### CODE REVIEW

File: pkg/validation/internal/multiarch.go
Comment (ryantking) [inline pkg/validation/internal/multiarch.go:95]: I think the following is more clear:

```go
if objMap, ok := obj.(map[string]string); ok && containerTool == “” {
    containerTool = objMap[ContainerToolsKey]
    break
}
```

File: pkg/validation/internal/multiarch.go
Comment (ryantking) [inline pkg/validation/internal/multiarch.go:102]: Similar comment to above:
```go
if bundle, ok := obj.(*manifests.Bundle); ok {
    // append bundle to results
}
```

File: pkg/validation/internal/multiarch.go
Comment (ryantking) [inline pkg/validation/internal/multiarch.go:125]: I think we should move this block up to `multipleArchitecturesValidate`. This function does not actually use `containerTool`, it just accepts it as a parameter and passes it into the validator. In my view, the function that retrieves the `containerTool` value should be the one to validate it and return an error.

File: pkg/validation/internal/multiarch.go
Comment (ryantking) [inline pkg/validation/internal/multiarch.go:132]: I think this section is a little bit confusing.

1. Why is `MultiArchValidator` exported when its primary function, `validate` is not?
2. Why are `Warns` read off the struct instead of being returned?

Could we do something like this?

```go
for _, err := range ValidateMultiArchBundle(bundle, containerTool) {
    // add the warn to the result
}

func ValidateMultiArchBundle(bundle *manifest.Bundle, containerTool string) []error {
    data := &multiArchValicatorData{
        AllArchFound: make(map[string]string),
        // etc
    }
    data.getInfraLabelsFromCSV(bundle)
    return data.checks() // this will return the warnings instead of mutating the struct
}
```

This gives us:

1. Single entrypoint to validate function, no instantiating struct then calling validate.

2. More self documenting since consumers of the API won’t need to know to read struct fields to get results. 

3. Don’t need to use single use struct.

File: pkg/validation/internal/multiarch.go
Comment (ryantking) [inline pkg/validation/internal/multiarch.go:139]:
```suggestion
	if len(containerTool) == 0 || strings.ToLower(containerTool) == "none" {
```

File: pkg/validation/internal/multiarch.go
Comment (ryantking) [inline pkg/validation/internal/multiarch.go:200]: In the case that `err` is `nil` and the length of the output is 0, you’ll get a weird silent failure. I would prefer:

```go
if err != nil {
    // return err
} else if len(output) == 0 {
    // return empty output error
}
```

The current code will return an empty manifest and nil error when the output is empty.

File: pkg/validation/internal/multiarch.go
Comment (ryantking) [inline pkg/validation/internal/multiarch.go:266]: What do we gain by not breaking here? The only thing that can happen is `found` gets set to `true` a second time.

File: pkg/validation/internal/multiarch.go
Comment (ryantking) [inline pkg/validation/internal/multiarch.go:337]:
```suggestion
		// If the value is linux and no labels were added to the CSV then it is fine
		if !found && !(supported == "linux" && len(data.InfraSOLabels) == 0) {
			notFoundSoLabel = append(notFoundSoLabel, supported)
		}
```

File: pkg/validation/internal/multiarch.go
Comment (ryantking) [inline pkg/validation/internal/multiarch.go:372]: Similar suggestion to above

**Review by ryantking — CHANGES_REQUESTED** [review]
Few changes and questions. A lot of it is stylistic stuff to make the code more readable.

File: pkg/validation/internal/multiarch.go
Comment (camilamacedo86) [reply pkg/validation/internal/multiarch.go:125]: The idea is valid for all input first. We could have other keys. 
So,  would not make sense to move forward without first checking if the input is correct. 
See that if we do what you suggest we will start to do the checks before ensuring that it is actually possible with the data provided.

File: pkg/validation/internal/multiarch.go
Comment (camilamacedo86) [reply pkg/validation/internal/multiarch.go:95]: Not sure if that would be more clear, what happens if we have more optional keys?

File: pkg/validation/internal/multiarch.go
Comment (camilamacedo86) [reply pkg/validation/internal/multiarch.go:102]: same here. If we have other types? 
I think I would prefer to stick with the current implementation which indeed is == across all validators.

File: pkg/validation/internal/multiarch.go
Comment (camilamacedo86) [reply pkg/validation/internal/multiarch.go:132]: done some improvements based on your feedback.

File: pkg/validation/internal/multiarch.go
Comment (camilamacedo86) [reply pkg/validation/internal/multiarch.go:139]: Good point. 
Changed on top for we ensure that we will have the lower case always.

File: pkg/validation/internal/multiarch.go
Comment (camilamacedo86) [reply pkg/validation/internal/multiarch.go:200]: For both, the return is always the same `return manifestInspect{}, err`.
Not sure if I could follow up on your suggestion.

File: pkg/validation/internal/multiarch.go
Comment (camilamacedo86) [reply pkg/validation/internal/multiarch.go:130]: So, I tried to clarify in the code and I am closing this one as solved let me know if you think that we should re-open.

File: pkg/validation/internal/multiarch.go
Comment (ryantking) [reply pkg/validation/internal/multiarch.go:211]: The situation I'm thinking of is when we have:
```go
output := ""
err := nil
```
The return value of `runManifestInspect` will be `ManifestInspect{}, nil`, which will be confusing since the caller will need to know the meaning of the `ManifestInspect` zero value. I think we would want `ManifestInspect{}, errors.New("no output returned from manifest tool")` to make it clear what happened.

File: pkg/validation/internal/multiarch.go
Comment (camilamacedo86) [reply pkg/validation/internal/multiarch.go]: It is totally possible.

File: pkg/validation/internal/multiarch.go
Comment (camilamacedo86) [reply pkg/validation/internal/multiarch.go:211]: The `no output returned from manifest tool` would return an error always in this case. 
How can we do a docker manifest inspect and the manifest be parsed and empty and no error returned?

Comment (ryantking) [reply pkg/validation/internal/multiarch.go:211]: Then why check for the empty output in the first place? From what you just said if `len(output) == 0` is `true` then `err != nil` will also be `true`.

I also don't want to hold up this PR on this relatively small edge case, so I am good to approve if you are confident that we need to check the output length and it will never be empty without an error, but I stand by that you should handle `len(output) == 0` in a separate `if/else` branch or drop the check entirely.

Comment (camilamacedo86) [reply pkg/validation/internal/multiarch.go]: I think you are right the len(output) == 0 is redundant.
Removed :-)

Comment (camilamacedo86) [reply pkg/validation/internal/multiarch.go]: I think you are right the len(output) == 0 is redundant.
Removed :-)

Comment (camilamacedo86) [reply pkg/validation/internal/multiarch.go]: I think you are right the len(output) == 0 is redundant.
Removed :-)

File: pkg/validation/internal/multiarch.go
Comment (jmrodri) [inline pkg/validation/internal/multiarch.go:104]: Why are we looping over the objs twice? Can we not do both checks in a single iteration?

**Review by jmrodri — CHANGES_REQUESTED** [review]
A couple suggestions and question about SO vs OS.

File: pkg/validation/internal/multiarch.go
Comment (camilamacedo86) [reply pkg/validation/internal/multiarch.go:54]: Really good catcher 👍 🥇 THANK YOU. 
All should be os + the label is "operatorframework.io/os"

File: pkg/validation/internal/multiarch.go
Comment (camilamacedo86) [reply pkg/validation/internal/multiarch.go:104]: If we put both together then can we not call the validate before getting the key?
So that is done first to get the key after we call the validate
Note that all are == across the board and I agree that we have a great hall for improvements.
However, wdyt about we let this for when we change one place we do == for all the other places?

File: pkg/validation/internal/multiarch.go
Comment (camilamacedo86) [reply pkg/validation/internal/multiarch.go:150]: See:

```
// Check the key values informed
			containerTool = obj.(map[string]string)[ContainerToolsKey]
			if len(containerTool) > 0 {
				// Make lower for we compare and use it
				containerTool = strings.ToLower(containerTool)
				break
			}
```
We ensure that the string is informed as the value of the key will be the lower case always so that here we just check if the value is or is not valid. Is that make sense?

File: pkg/validation/internal/multiarch.go
Comment (camilamacedo86) [reply pkg/validation/internal/multiarch.go:334]: should all be so right? :-) Typo issue

**Comment by camilamacedo86** [comment]
@jmrodri THANK YOU a lot.
All addressed, please feel free to review it again.

File: pkg/validation/internal/multiarch.go
Comment (camilamacedo86) [reply pkg/validation/internal/multiarch.go:177]: That was a good catcher. But the label is `Labels map[string]string`.
So that do not panic if be nil. See: https://go.dev/play/p/VUZPduMAOIf

File: pkg/validation/internal/multiarch.go
Comment (camilamacedo86) [reply pkg/validation/internal/multiarch.go:191]: since that I small nit if we have not other ones and if you do not mind I prefer to address in a follow up for we do lost the lgtm label :-)

**Review by ryantking — APPROVED** [review]

File: pkg/validation/internal/multiarch.go
Comment (perdasilva) [inline pkg/validation/internal/multiarch.go:18]:
```suggestion
// This validator will inspect the images with the chosen container-tool. One of: [docker, podman, none] (By default docker)
```

File: pkg/validation/internal/multiarch.go
Comment (bentito) [inline pkg/validation/internal/multiarch.go:21]:
```suggestion
// - raise a error(s) when is possible to confirm that images do not provide the support defined via to the labels on the CSV
```

File: pkg/validation/internal/multiarch.go
Comment (bentito) [inline pkg/validation/internal/multiarch.go:33]:
```suggestion
// - If your Operator bundle specifies images which do not supports all archetypes found for your Operator image(s) (probably supported by your project)
```

File: pkg/validation/internal/multiarch.go
Comment (bentito) [inline pkg/validation/internal/multiarch.go:35]:
```suggestion
// Note: To better guess the case scenarios where authors might have missed the labels, the following check will verify all architectures supported by the Operator image(s). However, by looking at the CSV we are not able to ensure what is the Operator image because this info is not provided. Therefore, we know by SDK the Operator image container will be called manager.
```

Comment (bentito) [reply pkg/validation/internal/multiarch.go:35]: I'm still not sure what is meant here?

File: pkg/validation/internal/multiarch.go
Comment (bentito) [inline pkg/validation/internal/multiarch.go:50]:
```suggestion
// operatorFrameworkOSLabel stores the labels for the supported OS from the CSV
```

File: pkg/validation/internal/multiarch.go
Comment (bentito) [inline pkg/validation/internal/multiarch.go:45]: Would `ContainerToolsPath` be a better name?

Comment (camilamacedo86) [reply pkg/validation/internal/multiarch.go:45]: I do not think so because it is a key, value map.
We have the same in other places so I would vote in keep all following the same standards.

Comment (bentito) [reply pkg/validation/internal/multiarch.go:45]: ok

File: pkg/validation/internal/multiarch.go
Comment (bentito) [inline pkg/validation/internal/multiarch.go:168]: But isn't "none" explicitly saying: "there is no container tool for this"? If that's the case, it should not ignore that setting and default to Docker.

Comment (camilamacedo86) [reply pkg/validation/internal/multiarch.go:168]: This is behaviour across the tools. (Keeping the same) None is == "empty"
Could you please check the commands in OPM and SDK that allows us to pass the container tools (i.e.: https://github.com/operator-framework/operator-sdk/blob/5f6c16c792b486c5d955a36b060112be288443ff/website/content/en/docs/olm-integration/cli-overview.md#private-bundle-and-catalog-image-registries)

Comment (bentito) [reply pkg/validation/internal/multiarch.go:168]: ah okay

**Review by dinhxuanvu — APPROVED** [review]
/approve

**Pull request merged**
